### ISSUE TITLE
Error: Message missing required keys.

### ISSUE THREAD

**Issue by jason-riddle** [aws/aws-js-sns-message-validator#2]
I'm trying to use this library with AWS Lambda to validate an incoming SNS message, but I keep running into the following error message.

```
Error: Message missing required keys.
```

After some more investigation, I've noticied a very subtle difference.  When logging what the SNS message looks like as it triggers the lambda and in AWS Lambda's SNS sample json event files, the SNS message looks like this (unimportant pieces snipped out)

``` JSON
{
      "Sns": {
        ...,
        "SigningCertUrl": "EXAMPLE",
        ...,
        "UnsubscribeUrl": "EXAMPLE",
        ...
}
```

While as documented here, http://docs.aws.amazon.com/sns/latest/dg/json-formats.html as well as the test cases you provide for this library, the json structure is the following.

``` JSON
{
      "Sns": {
        ...,
        "SigningCertURL": "EXAMPLE",
        ...,
        "UnsubscribeURL": "EXAMPLE",
        ...
}
```

Notice the subtle capitalization of the 'r' and the 'l' (U**rl** and the U**RL**). I'm not sure if this is a bug in the library itself or if AWS's documentation needs to be updated?

**Comment by jeskew** [comment]
Hi @jason-riddle,

The message formats differ in more than just capitalization. It looks like the Lambda event puts all SNS data in an object under the `Sns` key, whereas HTTP endpoint messages have `SigningCertURL` and the rest of the data points mentioned in [JSON format documentation](http://docs.aws.amazon.com/sns/latest/dg/json-formats.html) as top-level keys in the JSON payload sent.

Are you using a JSON generator in Lambda or are the issues just being caused by the sample JSON events you mentioned?

**Comment by jason-riddle** [comment]
The issues are being caused by the sample JSON events.

Here is the full sample SNS event template for Lambda.

``` JSON
{
  "Records": [
    {
      "EventVersion": "1.0",
      "EventSubscriptionArn": "arn:aws:sns:EXAMPLE",
      "EventSource": "aws:sns",
      "Sns": {
        "SignatureVersion": "1",
        "Timestamp": "1970-01-01T00:00:00.000Z",
        "Signature": "EXAMPLE",
        "SigningCertUrl": "EXAMPLE",
        "MessageId": "95df01b4-ee98-5cb9-9903-4c221d41eb5e",
        "Message": "Hello from SNS!",
        "MessageAttributes": {
          "Test": {
            "Type": "String",
            "Value": "TestString"
          },
          "TestBinary": {
            "Type": "Binary",
            "Value": "TestBinary"
          }
        },
        "Type": "Notification",
        "UnsubscribeUrl": "EXAMPLE",
        "TopicArn": "arn:aws:sns:EXAMPLE",
        "Subject": "TestInvoke"
      }
    }
  ]
}
```

Also, I have a github webhook that fires an SNS event, on any event, which looks like the following.

``` JSON
{
    "Records": [
        {
            "EventSource": "aws:sns",
            "EventVersion": "1.0",
            "EventSubscriptionArn": "arn:aws:sns:us-east-1:<SNIPPED>:github:<SNIPPED>",
            "Sns": {
                "Type": "Notification",
                "MessageId": "<SNIPPED>",
                "TopicArn": "arn:aws:sns:us-east-1:<SNIPPED>:github",
                "Subject": null,
                "Message": "<SNIPPED>",
                "Timestamp": "2015-11-28T22:08:19.167Z",
                "SignatureVersion": "1",
                "Signature": "<SNIPPED>",
                "SigningCertUrl": "<SNIPPED>",
                "UnsubscribeUrl": "<SNIPPED>",
                "MessageAttributes": {
                    "X-Github-Event": {
                        "Type": "String",
                        "Value": "push"
                    }
                }
            }
        }
    ]
}
```

In both of these cases, I'm simply just printing out the event received with the following.

``` javascript
JSON.stringify(event, null, 2);
```

**Comment by jeskew** [comment]
Hi again,

Just wanted to give you an update to let you know this issue hasn't been forgotten. I'm still working with the SNS team to find a backwards-compatible way to fix this.

**Comment by jeskew** [comment]
@jason-riddle While you're waiting on a more comprehensive fix, I found a temporary solution. The message is signed based on the published format, not on the Lambda format, so you can transform the message object before passing it to the validator:

``` javascript
var normalizeLambdaMesage = function (message) {
    var inconsistentKeys = {
        SigningCertUrl: 'SigningCertURL',
        SubscribeUrl: 'SubscribeURL',
        UnsubscribeUrl: 'UnsubscribeURL'
    };

    for (var lambdaKey in inconsistentKeys) {
        if (inconsistentKeys.hasOwnProperty(lambdaKey) && message.hasOwnProperty(lambdaKey)) {
            message[inconsistentKeys[lambdaKey]] = message[lambdaKey];
            delete message[lambdaKey];
        }
    }

    return message;
};
```

The resulting message can be validated in the same way as a regular SNS HTTP(S) notification.

**Comment by jason-riddle** [comment]
Cool. Thanks for the quick fix!

**Comment by rafaljanicki** [comment]
Hey,

Do you have any resolution for this error?

**Comment by jason-riddle** [comment]
So since https://github.com/aws/aws-js-sns-message-validator/pull/10 has been merged, this can be closed correct?

**Comment by jeskew** [comment]
It can! I tagged and published a new release with the changes from #10.

**Comment by emaincourt** [comment]
Hello !

I only wanted to advise you that I am still facing a related issue. It seems that you need to check the certificate URL with the right key before request. I get the following error :
```
{
    "errorMessage": "connect ECONNREFUSED 127.0.0.1:443",
    "errorType": "Error",
    "stackTrace": [
        "Object.exports._errnoException (util.js:1018:11)",
        "exports._exceptionWithHostPort (util.js:1041:20)",
        "TCPConnectWrap.afterConnect [as oncomplete] (net.js:1090:14)"
    ]
}
```

I fixed it this way :
```
const certUrl = message.SigningCertURL || message.SigningCertUrl;
getCertificate(certUrl, function (err, certificate) {});
```

Hope this might be helpful.
